container/openejb-core/src/main/java/org/apache/openejb/cdi/CdiBuilder.java

```diff
@@ -37,12 +37,14 @@
     private static final Logger logger = Logger.getInstance(LogCategory.OPENEJB_CDI, CdiBuilder.class);
     private static final ThreadSingletonService SINGLETON_SERVICE = new ThreadSingletonServiceImpl();
 
+    private static final RequestScopedThreadContextListener REQUEST_SCOPED_THREAD_CONTEXT_LISTENER = new RequestScopedThreadContextListener();
+
     public void build(final AppInfo appInfo, final AppContext appContext, final List<BeanContext> allDeployments) {
         initSingleton().initialize(new StartupObject(appContext, appInfo, allDeployments));
     }
 
     private ThreadSingletonService initSingleton() {
-        ThreadContext.addThreadContextListener(new RequestScopedThreadContextListener());
+        ThreadContext.addThreadContextListener(REQUEST_SCOPED_THREAD_CONTEXT_LISTENER);
```

Comment:
I'm not comfortable with this change as this serves request scoped beans, hence multiple threads and multiple applications.

Reply:
There is no need of multiple instances. RequestScopedThreadContextListener creates the scope and keeps it inside the supplied ThreadContext. 
Also initSingleton is called only once when CDI container is initialized. That container is initialized when first CDI enabled app is deployed. 

When new  instance is created when server is stopped and again started(multiple starts of tests in arquilian) new instance of RequestScopedThreadContextListener  will be added in the original code. Instead now same instance will be "added" to the set of listeners.

This listener is never removed from listener list/set

Reply:
ok